virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,317 @@
+
+1. extra-terrestrial radiation (affected by the earth's orbit, date, and location),
+2. terrestrial radiation (affected by atmospheric composition and clouds),
+3. topography (elevation, slope and aspect),
+4. surface albedo (vegetation type and fraction of vegetation/bare soil), and
+5. emitted longwave radiation.
+
+The preprocessing module takes extra-terrestrial radiation as an input and adjusts for
+the effects of topography (slope and aspect). Here, the effects of atmospheric
+filtering (elevation-dependent) and cloud cover are added to calculate photosynthetic
+photon flux density (PPFD) at the top of the canopy which is a crucial input to the
+plant module. The implementation is based on :cite:t:`Davis2017`.
+
+Cloud cover and surface albedo also determine how much of the shortwave radiation that
+reaches the top of the canopy is reflected and how much remains to be absorbed via
+photosynthesis and re-emitted as longwave radiation by vegetation and forest floor.
+At this stage, scattering and re-absorption of longwave radiation are not considered.
+
+"""
+# the following structural components are not implemented yet
+# TODO include time dimension
+# TODO logging, raise errors
+
+
+from dataclasses import dataclass, field
+from typing import Any, Dict
+
+import numpy as np
+from numpy.typing import NDArray
+
+from virtual_rainforest.core.logger import LOGGER
+from virtual_rainforest.core.model import InitialisationError
+
+# from some external source, could be core.constants or abiotic.config
+CLOUDY_TRANSMISSIVITY = 0.25
+"""Cloudy transmittivity :cite:p:`Linacre1968`"""
+TRANSMISSIVITY_COEFFICIENT = 0.50
+"""Angular coefficient of transmittivity :cite:p:`Linacre1968`"""
+FLUX_TO_ENERGY = 2.04
+"""From flux to energy conversion, umol J-1 :cite:p:`Meek1984`"""
+STEFAN_BOLTZMANN_CONSTANT = 5.67e-8
+"""Stefan-Boltzmann constant W m-2 K-4"""
+SOIL_EMISSIVITY = 0.95
+"""Soil emissivity, default for tropical rainforest"""
+CANOPY_EMISSIVITY = 0.95
+"""Canopy emissivity, default for tropical rainforest"""
+BEER_REGRESSION = 2.67e-5
+"""Parameter in equation for atmospheric transmissivity based on regression of Beer's
+radiation extinction function :cite:p:`Allen1996`"""
+CELSIUS_TO_KELVIN = 273.15
+"""Factor to convert temperature in Celsius to absolute temperature in Kelvin"""
+
+# dummy data object, 2 grid cells, 3 canopy layers
+data: Dict[str, Any] = {
+    "elevation": np.array([100, 100], dtype=np.float32),
+    "sunshine_fraction": np.array([0.5, 1.0], dtype=np.float32),
+    "shortwave_in": np.array([100, 100], dtype=np.float32),
+    "canopy_temperature": np.array(
+        [
+            [25, 22, 20],
+            [25, 22, 20],
+        ],
+        dtype=np.float32,
+    ),
+    "surface_temperature": np.array([20, 20], dtype=np.float32),
+    "canopy_absorption": np.array(
+        [
+            [300, 200, 100],
+            [300, 200, 100],
+        ],
+        dtype=np.float32,
+    ),
+}
+
+
+@dataclass
+class Radiation:
```

Comment:
Given that all but one of these fields is initialised in `__post_init__` (and that one seems to be a source of test failures), why not just make this a regular class and do the necessary checks/calculations in `__init__`? I don't see what we get by making this a `dataclass`, except weirder-looking code.

@davidorme, I think this was your idea, right?

Reply:
Hey! We all said it might work: https://github.com/ImperialCollegeLondon/virtual_rainforest/pull/155#discussion_r1094218746

I see what you mean about the 'weirder-looking' code though.

@vgro - I think we're getting into the realms of "equal functionality / flip flopping advice / implementation ping pong" here.

I think this is a good opportunity for pair programming because then you can have those "what the _hell_ are you talking about" discussions in real time, rather than working in a vacuum via GitHub commits.

Reply:
I agree, there are some very basic structural problems that I find hard to solve which leads to just more different ways of doing it wrong, a direct conversation would be very useful.

Reply:
I'd forgotten that :laughing:. Sorry, I wasn't trying to accuse you!

I did like the general idea of removing intermediate values from the class, but I'm not sure a `dataclass` is the way to go here (despite what I may have said before :wink:). As @davidorme says, there are a bunch of ways you could rewrite this, but I think it should be relatively easy to convert this into a regular class, with all the `__post_init__` stuff in `__init__` instead, which will just make it a bit clearer. Does anyone else have a better idea?